### PR TITLE
Pull in branch - Rak/resolve issue #41 per class metrics and #47 global metrics

### PR DESCRIPTION
**Pull request by rak5216**
Down to polishing and final details. See issue #41 and #47

### CODE REVIEW

**Comment by rak5216** [comment]
image

**Comment by rak5216** [comment]
image

**Comment by rak5216** [comment]
@Josh-Joseph major step before approval is to review per class metric/loss implementation and determine if other metrics should also be included (eg, f1-score, precision, recall)

**Comment by rak5216** [comment]
batch metrics written to new line issue: https://stackoverflow.com/questions/41442276/keras-verbose-training-progress-bar-writing-a-new-line-on-each-batch-issue

`import ipykernel` solved issue. There were too many metrics to fit on one prompt line, so new line printing triggered, even for `verbose=1` in `fit_generator`

image

**Comment by rak5216** [comment]
Global metrics (incorrect if computed per batch): precision, recall, f score (& dice loss), IOU (& jaccard loss)
Batch-capable metrics (correct for global with running avg of batch): accuracy

https://datascience.stackexchange.com/questions/45165/how-to-get-accuracy-f1-precision-and-recall-for-a-keras-model

**Comment by rak5216** [comment]
@Josh-Joseph (all classes in top row) Note quantitative differences with keras cat_ce vs. segmentation_models cat_ce, looks okay qualitatively. Also note differences b/w 3 accuracy types from keras. See #41 for equations details.

image

**Comment by Josh-Joseph** [comment]
hmm interesting! While it's a bit unsettling keras and sm values are different I don't think it's an issue for what we're using it for. And just defaulting then to then to the keras one makes sense to me.

**Comment by rak5216** [comment]
> hmm interesting! While it's a bit unsettling keras and sm values are different I don't think it's an issue for what we're using it for. And just defaulting then to then to the keras one makes sense to me.

@Josh-Joseph , for same dataset in train & val, the train loss is different than the val loss. I'll post plots and numbers when it's done soon

**Comment by rak5216** [comment]
image

**Comment by rak5216** [comment]
@Josh-Joseph, the one hot implementation in `models.py` is ready for a look. I'll report some results tables and plots here soon. The implementation is meant to be as simple as possible. I had a big issue initially with instatiating ` class OneHotMetricWrapper(object)` becasue the keras code needed to access `object.__name__`, which the instance doesn't have default access to (if i could change the keras code in `saving.py`, then `object.__class__.__name__` would work). Since I can't change the code, to give the class instance access to the class name, after much tinkering, I found one solution is to inherit the `KerasObject` base class from `segmentation_models`, which is well suited to the implementation of my wrapper class too. Then the key line to make things work with `class OneHotMetricWrapper(KerasObject):` is `super().__init__(name=self.name)`. This form is how both keras and _seg_models metric & loss classes plug-in easily to the keras framework.

File: models.py [hunk 1]
```diff
@@ -1,8 +1,31 @@
...
+from segmentation_models.base import functional
+
+SMOOTH = 1e-5
+
+
+class OneHotMetricWrapper(KerasObject):
+    def __init__(
+            self,
+            name=None,
+            metric_class_instance=None,   # DO I NEED A COMMA HERE?
```
Comment (Josh-Joseph) [inline models.py]: no you don't

**Comment by Josh-Joseph** [comment]
re val ratio not being 3 -- this is confusing and 100% a thing we should get to the bottom of. Yeah it makes sense it's not exactly the same as train but it should still be close. It is definitely starting to make me slightly concerned there's a bug somewhere. Maybe let's first solidify this one hot metric wrapper stuff and hopefully (but unlikely) it'll surface the issue. If it doesn't we can pick this back up then.

re one hot metric wrapper -- very cool you made that all work! One issue here is (imaging us 9 months from now or a new student who gets involved) many of these metrics computed batch wise will be wrong (the same reason keras pulled these out). I think assuming we'll remember or somehow pass on that information in documentation is not a great solution. So I think two choices here are to somehow throw extremely obvious warning when used incorrectly or (and my preferred route) is to only compute them when they are correct (i.e. only during validation and when it only has 1 batch). Thoughts?

**Comment by rak5216** [comment]
> re val ratio not being 3 -- this is confusing and 100% a thing we should get to the bottom of. Yeah it makes sense it's not exactly the same as train but I'd guess it'd still be relatively close. It is definitely starting to make me slightly concerned there's a bug somewhere. Maybe let's first solidify this one hot metric wrapper stuff and hopefully (but unlikely) it'll surface the issue. If it doesn't we can pick this back up then.

@Josh-Joseph , I'll look into this today and see if I can find any discrepancies. I agree they should be close in loss magnitude, and the proportionality b/w the two losses should be the same in train & val. I think we should confirm there is no bug before the big runs that are coming very soon (this week).

> re one hot metric wrapper -- very cool you made that all work! One issue here is (imaging us 9 months from now or a new student who gets involved) many of these metrics computed batch wise will be wrong (the same reason keras pulled these out). I think assuming we'll remember or somehow pass on that information in documentation is not a great solution. So I think two choices here are to somehow throw extremely obvious warning when used incorrectly or (and my preferred route) is to only compute them when they are correct (i.e. only during validation and when it only has 1 batch). Thoughts?

@Josh-Joseph , totally agree with this concern of it not being readily understood by other users. I believe evaluation (for all metrics except accuracy and CE) during val carries the same issue as eval during training, which is it needs to be done in batches due to varying denominators in the metric definitions (eg, IOU, precision, recall, f score) across batches (accuracy denominator is always # pixels = total # predicitions). HOWEVER, from what I just gathered, this issue isn't actually relevant here because none of the `segmentation_models` metrics get folded into rolling averages. So really, the s_m metrics are computed for a batch, then overwritten by the next batch's metric, such that the epoch-level metric is not an avg of all batches (which would be incorrect as we noted) but instead is the score of the final batch for that epoch. So, in this way, there is nothing incorrect about including these non-global metrics because they're not reflecting an epoch mean; I think they could be useful as "checkpoints" that assess each batch individually and then csvlog and plot the final batch's score. So the output s_m scores correctly show the performance change but only on the final batch, and the question becomes is that final batch of 8 or 16 images representative of the dataset or at least sufficiently representative such that it's at all beneficial to assess the model. Note, the built-in s_m computation strategy contrasts that of keras which emphasizes global measures. Keras metrics and losses accomplish global scope with a `MeanMetricWrapper` class that updates the score as rolling avg as each new batch is considered. Again, s_m is entirely missing this wrapper and reports metric for the current batch. (Though, for losses, s_m appears to record a rolling sum of batch means by inheriting a `Loss` super class.) In light of all of this, if we decide it's useful (even to the smallest degree) to record final batch scores with s_m metrics and global epoch scores with keras metrics/losses and s_m losses, then I can annotate the code more to be clear what score has what scope and also include it in user-define metric names

**Comment by rak5216** [comment]
@Josh-Joseph regarding differences in keras/s_m loss ratio fro training vs. val, here are some results that give further insight. The 1st (`loss`) and 3rd ('categorical_crossentropy`) items circled are identical functions from Keras, the 2nd (`sm_cat_cross_entropy`) is from segmentation_models. As expected, training shows 1st == 3rd == triple the 2nd. However in validation, 1st != 3rd now, even though they're defined identically. The difference is that the 1st compiled as a loss and the 3rd is compiled as a metric. So for some reason, in validation, loss functions are getting different data than metric functions.

image

**Comment by Josh-Joseph** [comment]
Let me see if I can summarize the current understanding of the facts floating around (and please feel free to edit my comment if I'm missing something/misunderstanding):

About metrics:
- There are two types of metrics, "global" and "local"
  - Global metrics require that they are computed across the entire dataset at once (eg across all of train or all of validation) and cannot be computed accurately if done in batches and then the metric values for each batch are averaged together. IOU score, F1 score, and Jaccard loss are examples of global metrics.
  - Local metrics can be computed on each batch and then the average of the metric value of each batch is equivalent to if the metric were to have been computed across the whole dataset at once. binary accuracy and categorical cross-entropy are examples of local metrics.
- Keras 2.0 removed all global metrics so any metric Keras supports is a local metric
- segmentation models has implementations of both global and local metrics and does not try to distinguish between the two
- some metrics, despite having the same name, have different implementations across keras and segmentation models (eg categorical cross entropy -- keras computes the sum of the entropy over the classes, segmentation models computes the average entropy over the classes)

About how Keras uses metrics:
- Keras has a concept of "stateful metrics". A stateful metric is one that is *not* averaged over batches. This means the stateful metric value will reflect the *last value of an epoch* -- so if we have multiple batches, it'll just be the value of the metric computed over the last batch (source [here](https://github.com/keras-team/keras/blob/master/keras/callbacks/callbacks.py#L477))
- Metrics given to Keras will be converted to a callback using BaseLogger as stateful metrics (see [here](https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/engine/training_generator.py#L87) and note that `model.metrics_names[0]` is `loss` created [here](https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/engine/training.py#L257))
- The loss function is not made a stateful metric

The core problem:
1. keras categorical cross entropy and segmentation models categorical cross entropy should produce values that are off by a factor of 3 (this is because we're running these over 3 classes and keras does a sum where segmentation models does a mean)
1. (1.) is true for train but not true for val *even when train and val have the same data*

Other weirdness:
- Even though the metrics computed during training are being computed against a model that is being continually updated through an epoch, we'd still expect the train and val curves to have roughly the same shape -- especially for when the train and val datasets are identical.

**Comment by Josh-Joseph** [comment]
If all this is true, I think our next test is to do a run with train and validate having epochs of 1 batch. This may be an issue where somehow (I have no idea how) different metrics are being computed for different batches?

**Comment by rak5216** [comment]
> @Josh-Joseph regarding differences in keras/s_m loss ratio from training vs. val, here are some results that give further insight. The 1st (`loss`) and 3rd (`categorical_crossentropy`) items circled are identical functions from Keras, the 2nd (`sm_cat_cross_entropy`) is from segmentation_models. As expected, training shows 1st == 3rd == triple the 2nd. However in validation, 1st != 3rd now even though we retain 3rd == triple the 2nd, again despite 1st and 3rd being called from Keras identically. The difference is that the 1st is compiled as a loss and the 3rd is compiled as a metric. So for some reason, in validation, loss functions are getting different data than metric functions.
> 
> I possibly found where `loss` is being treated differently during val vs. other metrics. [Conversion via float() vs. np.float64() messing it up?](https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/engine/training_generator.py#L420) (AND HERE IS GITHUB COMMIT W/O EXPLANATION - [`TRAINING_GENERATOR.PY` L393](https://github.com/keras-team/keras/commit/b938b91b527760cef3af29031e74ab181f2fc2b9#diff-b25d82c3f751f73f6e62b8455547ac73) ) --> [Possible answer?](https://stackoverflow.com/questions/27098529/numpy-float64-vs-python-float) --- 3 LINKS HERE
> 
> image

@Josh-Joseph, by the looks of this data output, `BaseLogger` seems to treat losses and metrics the same since the 1st (the loss) and 3rd (loss called as a metric) are both the same keras cat CE. So if loss is considered stateless, then it seems the metric list is considered stateless also. Despite what the source code seems to indicate...

**Comment by rak5216** [comment]
Model run: identical datasets for train, val, and test. 1 batch per epoch.

**Train + Val results:** ratio of keras loss to s_m loss = 3 for both training and val.

image

image

**Comment by rak5216** [comment]
**Model run:** identical datasets for train, val, and test. **10 batches per epoch.**

**Train + Val results:** ratio of keras loss to s_m loss != 3 for both training and val.

-

image


**Test results:** ratio of keras loss to s_m loss != 3.

-

**Comment by rak5216** [comment]
**Model run IV:** identical datasets for train, val, and test. **358 batches per epoch (max).**  REMOVED BASELOGGER FROM CALLBACK LIST

**Train + Val results:** ratio of keras loss to s_m loss

**Comment by rak5216** [comment]
**Model run V:** identical datasets for train, val, and test. **358 batches per epoch (max).**  NOTHING DECLARED STATEFUL.

```    
results = compiled_model.fit_generator(
        train_generator,
        steps_per_epoch=len(train_generator),
        epochs=epochs,
        validation_data=validation_generator,
        validation_steps=len(validation_generator),
        # callbacks=[BaseLogger(stateful_metrics=['keras_categ_ce_stateful', 'categ_acc_stateful']), model_checkpoint_callback, tensorboard_callback, tensorboard_image_callback, csv_logger_callback]
        callbacks=[BaseLogger(stateful_metrics=None), model_checkpoint_callback, tensorboard_callback, tensorboard_image_callback, csv_logger_callback]
        # callbacks=[model_checkpoint_callback, tensorboard_callback, tensorboard_image_callback, csv_logger_callback]
 )
```

**Train + Val results:** ratio of keras loss to s_m loss:




**Test results:** ratio of keras loss to s_m loss ?= 3.

**Comment by rak5216** [comment]
@Josh-Joseph seems like this person destroyed the ability to define stateful metrics:
https://github.com/keras-team/keras/commit/87b2b00c976d26451f149d7365ee194912930cde#diff-b25d82c3f751f73f6e62b8455547ac73

image

image

**Comment by rak5216** [comment]
@Josh-Joseph also as you can see in `training_generator.py` loss is auto declared not stateful but val_loss is auto declared stateful

image

**Comment by Josh-Joseph** [comment]
Nice digging! Couple comments:
(1)
> @Josh-Joseph seems like this person destroyed the ability to define stateful metrics:
From a bit of googling -- that contributor is a software engineering at google, so it's safe to assume these changes were deliberate. Generally there'd be an issue/doc somewhere describing this new "metric framework" but I can't seem to find it.

I'm not at my work machine at the moment so I can't quite check but another thing we should look into is that while those changes have be committed to the Keras master branch, they may not yet be included into the pypi keras package that is installed using `pip` (e.g. here: https://pypi.org/project/Keras/). So there may be a discrepancy between what we're seeing on the bleeding-edge keras master branch in github and what is installed/used on our machines.


(2)
> loss is auto declared not stateful but val_loss is auto declared stateful
Awesome. This is a huge find I think and may explain a lot! I assume this also means the way we use an evaluation generator for test means test is also then declared stateful?

>Weird since loss has the ratio=3 but val_loss does not have ratio=3 though.
I think this actually makes sense though (if I'm following correctly). I think the ratio=3 will not happen when things are stateful since the metrics are not being computed correctly in stateful mode?


(3)
> The += I think is an extend list operation in this case
Yep that's correct

**Comment by rak5216** [comment]
> Nice digging! Couple comments:
> (1)
> 
> > @Josh-Joseph seems like this person destroyed the ability to define stateful metrics:
> 
> From a bit of googling -- that contributor is a software engineering at google, so it's safe to assume these changes were deliberate. Generally there'd be an issue/doc somewhere describing this new "metric framework" but I can't seem to find it.
> 
> I'm not at my work machine at the moment so I can't quite check but another thing we should look into is that while those changes have be committed to the Keras master branch, they may not yet be included into the pypi keras package that is installed using `pip` (e.g. here: https://pypi.org/project/Keras/). So there may be a discrepancy between what we're seeing on the bleeding-edge keras master branch in github and what is installed/used on our machines.
> 

Looked at the pypi keras package (version 2.3.1, Oct. 7, 2019) and it seems to align with what I've seen on the github repos, which makes sense since the corresponding master merge was done on Sept. 9, 2019.

**Comment by rak5216** [comment]
> Weird since loss has the ratio=3 but val_loss does not have ratio=3 though.
> 
> I think this actually makes sense though (if I'm following correctly). I think the ratio=3 will not happen when things are stateful since the metrics are not being computed correctly in stateful mode?

I think it seems weird because `loss` should be the metric that behaves differently, since it's the only non-stateful metric. If `loss` is not stateful and the `s_m_loss` is stateful (along with all other metrics), how is it possible that state / non_stateful exhibit the ratio of 3?

**Comment by rak5216** [comment]
> Awesome. This is a huge find I think and may explain a lot! I assume this also means the way we use an evaluation generator for test means test is also then declared stateful?

One would think, but for the 1 batch epochs (model runs I and II above), i got different test ratios vs. val/training...

**Comment by rak5216** [comment]
@Josh-Joseph just to keep the pts where i'm stuck flowing, the following pts (out of many) stick out in my mind:

1. `BaseLogger` and `ProgbarLogger` are automatically included as part of `fit_generator` inside `training_generator.py`. Both of these callbacks accept stateful metrics, and should be called together as arguments then if stateful metrics are to be included. By default, we see everything except `loss` is defined as stateful; however, training `loss` has the same value as `keras_loss`, which is just cat_ce called again but as metric. 

- [ ] How can this be if `loss` is stateful and `keras_loss` is not? Is there somewhere down the pipe where the stateful metrics default is being overwritten, such that the `None` argument value for `stateful_metrics` actually means no stateful_metrics AND NOT metrics_names [1: ] (current meaning)? I looked into this and couldn't find any obvious overwrite down the

**Comment by rak5216** [comment]
just can't see the bigger picture here, unless the actual intent is assume everything is stateful. This assumption is especially strange because all global metrics were purged from keras...which is completely contradictory. Stateful is global of a batch. July 6, 2019 commit in `training_generator.py`:
image

**Comment by rak5216** [comment]
@Josh-Joseph might have gotten some more insight based on [keras-team/keras#9253](keras-team/keras#9253). This reverses again my statement from 7 days ago about what code actually computes the epoch mean.

> Initial thoughts on metric/loss outputs. NOW FOUND TO BE WRONG, DUE TO KERAS class BaseLogger(Callback): However, from what I just gathered, this issue isn't actually relevant here because none of the segmentation_models metrics get folded into rolling averages. So really, the s_m metrics are computed for a batch, then overwritten by the next batch's metric, such that the epoch-level metric is not an avg of all batches (which would be incorrect as we noted) but instead is the score of the final batch for that epoch. So, in this way, there is nothing incorrect about including these non-global metrics because they're not reflecting an epoch mean; I think they could be useful as "checkpoints" that assess each batch individually and then csvlog and plot the final batch's score. So the output s_m scores correctly show the performance change but only on the final batch, and the question becomes is that final batch of 8 or 16 images representative of the dataset or at least sufficiently representative such that it's at all beneficial to assess the model. Note, the built-in s_m computation strategy contrasts that of keras which emphasizes global measures. Keras metrics and losses accomplish global scope with a MeanMetricWrapper class that updates the score as rolling avg as each new batch is considered. Again, s_m is entirely missing this wrapper and reports metric for the current batch. (Though, for losses, s_m appears to record a rolling sum of batch means by inheriting a Loss super class.)
> 
> UPDATE: I AM WRONG ON HOW THE RUNNING AVG IS CONTROLLED. Instead of MeanMetricWrapper initiating epoch averaging for Keras metrics/losses, it seems epoch averaging is controlled for all metrics (both Keras and s_m) at a higher level in fit_generator via class BaseLogger(Callback) . Further, it looks like we can turn off epoch averaging for arbitrary metrics to avoid incorrect epoch averaging and target on_batch_end metrics. I think this global vs. not-global is identical to stateful vs. stateless metrics, which I haven't grasped hitherto.

Still working this out in my head. It seems the assumption that all metrics are stateful (ie, [1: ]) is deliberate and based on the wrapping of all keras metrics with the `MeanMetricWrapper`. Without the stateful assumption, basically there should be double averaging over a batch: once by BaseLogger and once again my the wrapper. Previously i thought this wrapper was superfluous due to the `BaseLogger` callback, in terms of who actually controls averaging. So anything coming from keras.metrics (which includes `class CategoricalCrossentropy(MeanMetricWrapper)`) has built-in batch-averaging over an epoch.  In contrast, anything coming from keras.losses (which includes `class CategoricalCrossentropy(LossFunctionWrapper)`) does not have built-in averaging, thus the averaging is done by `BaseLogger` via [1: ]. Interestingly, keras provides all loss functions in keras.losses as metrics in keras.metrics. At first glance, everything seems identical, and Keras says:

> A metric function is similar to a loss function, except that the results from evaluating a metric are not used when training the model. You may use any of the loss functions as a metric function.

The key detail not explained here is that batch-averaging is enabled automatically if call from keras.metrics.

An interesting point in our code is that for setting the loss argument inside `compile`, we use `from keras.metrics import binary_crossentropy, categorical_crossentropy`, which seems at first glance like we're committing the double-avg sin by feeding a mean-metric-wrapped fcn to a stateless argument inside `BaseLogger`. Not the case actually, since these two loss functions are featured in `keras.metrics` as `from .losses import categorical_crossentropy`. So we're actually indirectly importing the function from `keras.losses`. This is identical to keras instructions on their website:
> from keras import losses

>model.compile(loss=losses.mean_squared_error, optimizer='sgd')

The only uncertainty for me surrounds the fact that a better way seems to be what's listed in their git repos:

>     Usage with the `compile` API:
    ```python
    model = keras.Model(inputs, outputs)
    model.compile('sgd', loss=keras.losses.CategoricalCrossentropy())

because then you inherit other features, which don't seem necessary during compilation I guess though, since the website doesn't show this option. Keras `compile` states all are fine:

> loss: String (name of objective function) or objective function or `Loss` instance. See losses. If the model has multiple outputs, you can use a different loss on each output by passing a dictionary or a list of losses. The loss value that will be minimized by the model will then be the sum of all individual losses.

So the above discussion resolves the question about stateful_metrics in `BaseLogger` and `ProgbarLogger`: assume all metrics are stateful to avoid double averaging since averaging is done with `MeanMetricWrapper`, and assume the loss is stateless because class and function definitions in `keras.losses` do not have built-in averaging via the `LossMetricWrapper`.

This also points to the issue with training `loss` (stateless) vs. `val_loss` (stateful). It seems `val_loss` is created as a copy of the `keras.loss` instance, but instead should be the corresponding version found in `keras.metrics`. Unfortunately, we cannot control this from outside the source code.

**Comment by rak5216** [comment]
hmmm, things aren't as clear as i thought. Perhaps `val_loss` is not made stateful via the definition in the keras repos code. Have to looker deeper

**Comment by Josh-Joseph** [comment]
So I've been trying to dig through the keras codebase and I *still* don't have a great answer to if/where/how `val_loss` is made/not made stateful...but, in the meantime, let me see if I can summarize some of this. *If*:
- `loss` is automatically made stateless and
- `val_loss` (and all other metrics in both train and validate) are automatically made stateful
- => that explains all the strange results we've seen where that train and val ratio don't line up

Despite not being able to confirm this in the code -- if that's true that those are automatically set that way is the conclusion correct?

**Comment by rak5216** [comment]
> So I've been trying to dig through the keras codebase and I _still_ don't have a great answer to if/where/how `val_loss` is made/not made stateful...but, in the meantime, let me see if I can summarize some of this. _If_:
> 
> * `loss` is automatically made stateless and
> * `val_loss` (and all other metrics in both train and validate) are automatically made stateful
> * => that explains all the strange results we've seen where that train and val ratio don't line up
> 
> Despite not being able to confirm this in the code -- if that's true that those are automatically set that way is the conclusion correct?

@Josh-Joseph I cannot find where `val` would be made stateful. Contrary to what i though before, in BaseLogger and ProgbarLogger callback settings in `training_generator.py`, stateful_metrics are defined as .metric_names[1: ], which actually does not include the `val_` names. So not sure if stateful/stateless is preserved in the `val_` copy of the metric or not yet.

**Comment by rak5216** [comment]
just updating that i ran some verification last night and I'm still not understanding a few things. One would be why callbacks treats val_loss as stateful, showing only the last batch value (contrary to what I thought, i was overlooking the fact the val metric names are not explicitly passed as stateful_metrics). Another thing is why I can't repeat the loss/val_loss discrepancy in stateful status. I passed keras.losses.CategoricalCrossentropy as a metric, which should not inherit MeanMetricWrapper class and is declared stateful, yet the training and val values match those of keras.metric.CategoricalCrossentropy as a metric, which does inherit MeanMetricWrapper . I was expecting keras.losses.CategoricalCrossentropy (inherits Loss and LossFunctionWrapper-which only averages over 1 batch not all batches) to show last batch but something is causing it to be averaged over all batches. I couldn't find any safety mechanism in Keras that would convert losses.CatCE to metrics.CatCE when passed as metric. When keras.losses.CategoricalCrossentropy is passed as the loss, epoch averaging is done in training by BaseLogger (only stateless metric); however, again, the val_loss is treated as stateful oddly.

**Comment by rak5216** [comment]
> just updating that i ran some verification last night and I'm still not understanding a few things. One would be why callbacks treats val_loss as stateful, showing only the last batch value (contrary to what I thought, i was overlooking the fact the val metric names are not explicitly passed as stateful_metrics). Another thing is why I can't repeat the loss/val_loss discrepancy in stateful status. I passed keras.losses.CategoricalCrossentropy as a metric, which should not inherit MeanMetricWrapper class and is declared stateful, yet the training and val values match those of keras.metric.CategoricalCrossentropy as a metric, which does inherit MeanMetricWrapper . I was expecting keras.losses.CategoricalCrossentropy (inherits Loss and LossFunctionWrapper-which only averages over 1 batch not all batches) to show last batch but something is causing it to be averaged over all batches. I couldn't find any safety mechanism in Keras that would convert losses.CatCE to metrics.CatCE when passed as metric. When keras.losses.CategoricalCrossentropy is passed as the loss, epoch averaging is done in training by BaseLogger (only stateless metric); however, again, the val_loss is treated as stateful oddly.

Think I found out why I can't achieve stateful loss values from defining metric=keras.losses.CatCe. Turns out that during compile, keras indeed converts metric=keras.losses.fcn to metric=keras.metrics.fcn in order to attain the `MeanMetricWrapper`, so it can't be tricked into giving last batch values unless i write a custom metric. 

The exact sequence responsible for the .losses to .metrics conversion is:

`training.py` : ` (L33) compile` -> ` (L211) self._cache_output_metric_attributes(metrics, weighted_metrics)`  -> `(L737) self._per_output_metrics = training_utils.collect_per_output_metric_info(...)`

`training_utils.py` : `(L945) # If the metric function is not stateful, we create a stateful version.
            if not isinstance(metric_fn, metrics_module.Metric):
                metric_fn = metrics_module.MeanMetricWrapper(
                    metric_fn, name=metric_name)`

**Comment by rak5216** [comment]
@Josh-Joseph, i think we can fully focus on the stateful aspect (required by global metrics). My most recent statement was just trying to get stateful capability for the sake of it, by circumventing `MeanMetricWrapper` auto conversion if the metric isn't an instance of `from keras.metrics import Metric`. But solving this problem isn't really so helpful because we ultimately don't want last batch values, we want global metrics---which invokes addressing issue #47 in parallel. This may involve returning to keras global metrics that actually allow per class capability, much to my surprise, like s_m, so per class metrics and correct global metrics (issues #41 and #47) are linked. This is because bottomline we need global metrics to not inherit `MeanMetricWrapper`, however, `Metric` is a metaclass with metamethods, so we can't outright instantiate sometheing that inherits without overriding the metamethods, which are `update_states` and results`. In fact, the global metrics buried in keras.metrics.py show how this can be done, since for example, we see `class Precision(Metric)`. Inside the Precision class, we see these two methods defined. Ideally, for IOU, precision, recall, and fscore (and jaccard and dice losses by extension), we want a running total of TP, TN, FP, and FN's, and then we can piece together the metrics at the end of each epoch, or val, or test. I'm not quite sure if this is how the global metrics currently operate or not, but we could custom write the two methods if needed.

**Comment by rak5216** [comment]
**note on global metrics:** keras-based precision, recall, fscore, and iou are not computing class avgs.

**Comment by rak5216** [comment]
**Model run VI:** 
- •	identical datasets for train, val, and test. 
- •	358 batches per epoch (max). 
- •	Finalize one hot, manually check metric instantiation memory location and inheritance chain
- •	understand callbacks (do not customize baselogger or progbarlogger), 
- •	understand stateful (inherits `keras.metrics.Metric`)/stateless (inherits `keras.metrics.MeanMetricWrapper`--default behavior if `Metric` not a parent class), 
- •	ignore `val_loss`, 
- •	rewrite `PerClassBinaryAccuracy`, 
- •	verify global metrics, know they’re not class-averaged but globally accumulated, and test threshold effects (0.5), 
- •	rewrite per class metric calls.
- •	verify val vs. test identity if same datasets
- •	added `metrics_utils.py` to cleanup structure.

```
results = compiled_model.fit_generator(
        train_generator,
        steps_per_epoch=len(train_generator),
        epochs=epochs,
        validation_data=validation_generator,
        validation_steps=len(validation_generator),
        callbacks=[model_checkpoint_callback, tensorboard_callback, tensorboard_image_callback, csv_logger_callback]
 )
```

all metrics_analysis


models_segmentation-model-small_2cropClass_3class_VarMinPx_final-train3X_20200124T154724Z_plots_metrics_mosaic


postcompile-metrics

**Comment by Josh-Joseph** [comment]
wow what a pull request this turned out to be :)

all seems good, feel free to merge/delete the branch!

**Comment by rak5216** [comment]
The issues #41 and #47 are now resolved. Some other notes will follow before closing this PR

**Comment by rak5216** [comment]
`val_loss` issue: `https://github.com/keras-team/keras/issues/13641`

corresponding code: `https://github.com/keras-team/keras/blob/master/keras/engine/training_generator.py#L420`

Somewhere in the code, `outs_per_batch` is told to store stateful for the first element (`val_loss`) and stateless for the other elements ('val_' + metrics). This narrows the focus a lot, but I haven't found it yet

**Comment by rak5216** [comment]
Also I discovered after looking into the auto stateful_metrics declaration that tf.keras and keras are similar but not identical, including for how this particular code segment is organized, tho tf.keras does something similar. Sometimes tho the two are out of sync, which some github commits will sometimes call out

**Comment by rak5216** [comment]
yeah the post-metric score averaging approach would be similar to the one hot approach, in which i would make a mean subclass. However, similar to one hot, i would need to write a unique mean wrapper subclass for each metric and also a subclass for the one-hot subclass of each metric, which each will inherit a given metric, since the inherited methods for each metric are different. I spent more than a day working on different methods to make a dynamically defined subclass definition that includes changing inheritance, via factory methods, nested classes, import as repeatedly, and the type function, but there is seemingly no simple answer in python for class (not instance) definition on the fly. `type` seemed close, but i couldn't control So i had to write out the one hot subclass definition for each relevant metric. Otherwise, trying to write just one subclass definition would involve changing inheritance depending on the metric i am trying to one hot, which changes the varying instance methods inherited from the parent class. Seems like classes can't be redefined during the run anyways, so I guess that eliminates worry about mixing up instance methods (@josh)? Thus by writing a separate subclass for each metric that inherits the metric, I can achieve manipulation of the pred and label data either going into or coming out of the parent class __call__ method, and all other parent methods are preserved, so seems makes a seamless integration in the keras machine. This is easier to explain in person or over the phone if needed before the next meeting.

With all that said, doing all of this to achieve averaging seems unnecessary, since it can more easily be done in postprocessing given the mix of imported (fixed) and custom metrics. I think consistency is best done here by just noting in the code comment that global metrics are not averaged (ie, scaled by num_classes) and need to be averaged after program completion by dividing by num_classes. 

I think the type(classname, superclasses, attributes_dict) function could be used to dynamically create subclass definitions that preserved all methods of metric except for the data manip to output mean (this could also be used to rewrite the one hot approach (same results just less code)). Two class levels will be added to each metric. Metric subclass is mean (or one-hot) level that I somehow write to be agnostic wrt inheritance (or use factory fcn or dynamic creation). Subclass inherits __init __ from parent (preservation) and only other subclass method is a tweak to __call __ to perform data manip then return parent call . I basically copy/pasted this format for each metric i wanted to one-hot, so that class name, inheritance, and init attributes only really changed.  Then the subsubclass would be created though type and inherit the different combinations of methods needed for each metric. I could run a loop to create what's needed based only on what's defined in a global metrics list...hmmm it would def be less readable tho since subsubclasses would be created at runtime. I think the main obstacles are I need to dynamically name the subclass and the subsubclass (type is the only way i know to do this), but if i use type, then i am not sure how to modify the newly created class's __call __ to be a mix of the data manip treatment + inherited call. Possibly with a fcn like
```
def func(cls, self, true, pred): 
    data manip
    return super(self.__class__, self).__call__()
```
I guess this thinking trajectory would be monkey patching
https://www.python-course.eu/python3_classes_and_type.php
http://alexgaudio.com/2011/10/07/dynamic-inheritance-python.html
https://stackoverflow.com/questions/38541015/how-to-monkey-patch-a-call-method

and with all that said, none of it is actually useful (unless we want to make one hot more concise/less code, which isn't needed). The global metrics only need averaging when they are first computed per class, followed by avg'ing to get to a single score. In our case, we're computing differently, all classes are blended globally (aka all simultaneously accumulated) together. Both cases still have a range from 0 to 1. In the posted results, you'll see the following differences between class-avg'd IoU and f1score and their global counterparts. I guess it's up to us to decide which is more useful. Either way, thankfully the mean subclass in unnecessary then.

**Comment by rak5216** [comment]
**Model run VII:** 
•	Added 1H1H and 1HUS and more metrics to verify 1H is working and try to understand accuracy
•	Based on comparisons with tf2 in Pull Request 48, 1H in tf1 did not work in most cases, except for when inheriting non-Keras classes and being converted to keras.MeanMetricWrapper
•	rerun test from Model Run VI
o	reopened this branch based ontf2 comparison (#48 ) showing that 1H was not working apparently in prior model runs in tf1 

```
pipenv run python3 test_segmentation_model.py --gcp-bucket gs://necstlab-sandbox --dataset-id dataset-small_2cropClass_3class_VarMinPx_final-train3X --model-id segmentation-model-small_2cropClass_3class_VarMinPx_final-train3X_20200124T154724Z
```

metrics

**Comment by rak5216** [comment]
CONCLUSION: TF1 NEEDS ONE HOT IMPLEMENTED AT UPDATE_STATE NOT __CALL__

**Pull request merged**